### DISCUSSION TITLE
Improve file management with new MKDIR command in RIA monitor.

### DISCUSSION THREAD

**Discussion by vruumllc** [picocomputer/community#76] [Q&A]
Hi, my usb stick root directory is getting pretty messy and there is currently no way to make subdirectories without removing my usb stick from my case and using a PC, which means unscrewing the case top, etc. 

I successfully added a MKDIR command to the RIA with the following mods:
![mkdir](https://github.com/picocomputer/community/assets/145504135/36ace86e-24a1-457c-b42a-1dcf33346347)

 Note that the existing UNLINK command will remove an empty directory as well as files.

With this change in place, I can now create my desired directory structure on the RP6502, change to the desired target directory, then run the python "upload" comand in rp6502.py from my computer to put new files in the desired location.

Addition of a copy and rename commands to the RIA monitor would also be welcome, but the existing command structure doesn't support the required two arguments. The addition of this MKDIR to the next release is easy, and is sufficient for my needs.

Thanks, tonyvr

OOPS! addendum: I guess the parentheses around the arguments in the help should not be there, as the arguments aren't optional.
![help](https://github.com/picocomputer/community/assets/145504135/32428de6-9168-4d01-8b6c-76be51d3a619)

Also, "HELP MKDIR" should print some meaningful help, and the help for UNLINK could be improved.

**Comment by vruumllc** [comment]
I saw this was checked in to main branch. Thanks!

**Comment by n9jcv** [comment]
Thank you @vruumllc   It is nice to see the community helping to improve our picocomputers.  Unfortunately, I do not have the skills to do these things.

**Reply by vruumllc** [reply]
Thanks! 
Stick with it and keep learning!  You solved your build problems, which shows you have the perseverance to succeed.

I highly recommend Ben Eater's video to learn about microprocessors and the 6502 specifically : 
https://eater.net/

Hunter Adam's Cornell course videos are an amazing intro to the rp pico and microcontroller projects in general.
https://www.youtube.com/playlist?list=PLDqMkB5cbBA5oDg8VXM110GKc-CmvUqEZ

**Reply by rumbledethumps** [reply]
>I do not have the skills to do these things.

An open source project needs all kinds of skills.

**Reply by n9jcv** [reply]
@vruumllc @rumbledethumps Yes i am in the middle of a ben eater 8 bit build now as well.  I am learning a lot from that.

If there is an opportunity to help with this project, reach out to me please.

**Comment by WojciechGw** [comment]
@vruumllc Thank You :)

> Addition of a copy and rename commands to the RIA monitor would also be welcome, but the existing command structure doesn't support the required two arguments. The addition of this MKDIR to the next release is easy, and is sufficient for my needs.

Perhaps using two arguments packed into one with a comma between them and then parsed to the two required by f_rename?
like
`]rename oldname.rp6502,newname.rp6502`
instead of
`]rename oldname.rp6502 newname.rp6502`

something like

fil.c
-----------------
```
#include <string.h>

void fil_mon_rename(const char *args, size_t len)
{
    FRESULT result;
    if (!len)
    {
        printf("?Missing arguments\n");
        return;
    }
    char* arg1 = strtok((char *)args, ",");
    char* arg2 = strtok(NULL, ",");
    result = f_rename(arg1, arg2);
    if (result != FR_OK)
        printf("?Unable to rename (%d)\n", result);
}
```

fil.h
-----------------
`void fil_mon_rename(const char *args, size_t len);`

hlp.c
-----------------
`    "RENAME name,newname - Rename a file or directory name.\n"`

```
static const char __in_flash("helptext") hlp_text_rename[] =
    "RENAME is used to rename a file or directory.";
```
`    {6, "rename", hlp_text_rename},`
		
mon.c
-----------------
`    {6, "rename", fil_mon_rename},`

**Comment by WojciechGw** [comment]
or different approach

command's has two arguments `RENAME filename newfilename`

I successfully added a RENAME command to the RIA with the following mods in proper places:

fil.c
----------------
```
void fil_mon_rename(const char *args, size_t len)
{
    char lfs_name[LFS_NAME_MAX + 1];
    char lfs_name_new[LFS_NAME_MAX + 1];
    FRESULT result;
    if (parse_file_name(&args, &len, lfs_name) &&
        parse_file_name(&args, &len, lfs_name_new) &&
        parse_end(args, len))
    {
        result = f_rename(lfs_name, lfs_name_new);
        if (result != FR_OK)
            printf("?Unable to rename (%d)\n", result);
    } else {
            printf("!invalid or missing argument(s)!\n");
    }
}
```

fil.h
-----------------
`void fil_mon_rename(const char *args, size_t len);`

hlp.c
-----------------
`    "RENAME name newname - Rename or move a file or directory within current drive.\n"`

```
static const char __in_flash("helptext") hlp_text_rename[] =
    "RENAME is used to rename or move a file or directory within current drive.\n"
    "usage examples:\n"
    "- rename file\n  ]RENAME info.txt info.old\n"
    "- move file to directory\n  ]RENAME info.txt archive/info.txt\n"
    "- move directory to parent directory\n  ]RENAME archive ../archive";
```

`    {6, "rename", hlp_text_rename},`

mon.c
-----------------
`    {6, "rename", fil_mon_rename},`
		
str.c
-----------------
```
bool parse_file_name(const char **args, size_t *len, char *name)
{
    name[0] = 0;
    size_t name_len = 0;
    size_t i;
    for (i = 0; i < *len; i++)
    {
        if ((*args)[i] != ' ')
            break;
    }
    if (i == *len)
        return false;
    for (; i < *len && name_len < LFS_NAME_MAX; i++)
    {
        char ch = (*args)[i];
        if (ch == ' ')
            break;
        if (((ch >= 'a' && ch <= 'z') 
            || (ch >= 'A' && ch <= 'Z'))
            || (ch == '.')
            || (ch == '/')
            || (ch == '_')
            || (name_len && ch >= '0' && ch <= '9'))
        {
            name[name_len++] = ch;
            continue;
        }
        name[0] = 0;
        return false;
    }
    if (!name_len)
        return false;
    if (i < *len && (*args)[i] != ' ')
    {
        name[0] = 0;
        return false;
    }
    for (; i < *len; i++)
    {
        if ((*args)[i] != ' ')
            break;
    }
    *len -= i;
    *args += i;
    name[name_len] = 0;
    return true;
}
```

str.h
--------------
```
// A file name
// Only A-Z a-z 0-9 and dot allowed
// Return argument name must hold LFS_NAME_MAX+1.
bool parse_file_name(const char **args, size_t *len, char *name);
```

**Reply by vruumllc** [reply]
EXCELLENT! I guess I was wrong about the existing command structure only supporting one arg. (That NEVER happens 8^)
I can't wait to try this out. Thanks

I would, however, prefer the linux 'mv' instead of rename

**Reply by WojciechGw** [reply]
You are very welcome :)

parse_file_name still needs to be refined
in terms of the characters allowed in the filename

**Comment by rumbledethumps** [comment]
This is not the direction I want things to go. These things belong on the 6502 side. The boot monitor is not a catch all, it's a monitor for booting the system.

**Reply by WojciechGw** [reply]
Yes I get it and respect, but - IMHO - possibility of rename is quit useful ...

**Reply by vruumllc** [reply]
Yes, I agree. The RIA already handles some of the basic file management. 
I think adding a CP and MV command to the monitor would be appropriate and sufficient.

**Comment by WojciechGw** [comment]
I've made some minor changes to fil.c, hlp.c and str.c (the above code has already been corrected).
It's now possible to use RENAME command also to move file or directory within current drive.

**Comment by rumbledethumps** [comment]
There's 2KB free on the RIA before memory gets complicated. I understand this seems easy and useful, but once that 2KB is gone then ALL future work becomes more difficult.

**Reply by WojciechGw** [reply]
I really understand the full extent of the problem - Pico memory isn't made of rubber ;)
so let's treat this as a fun "I can do it" game :)
